garage/np/algos/batch_polopt.py

```diff
@@ -1,135 +1,84 @@
-                samples_data = self.sampler.process_samples(itr, paths)
-                self.log_diagnostics(paths)
-                self.optimize_policy(itr, samples_data)
-                logger.log('Saving snapshot...')
-                params = self.get_itr_snapshot(itr, samples_data)
-                self.current_itr = itr + 1
-                params['algo'] = self
-                if self.store_paths:
-                    params['paths'] = samples_data['paths']
-                snapshotter.save_itr_params(itr, params)
-                logger.log('saved')
-                logger.log(tabular)
-                if self.plot:
-                    plotter.update_plot(self.policy, self.max_path_length)
-                    if self.pause_for_plot:
-                        input('Plotting evaluation run: Press Enter to '
-                              'continue...')
-
-        plotter.close()
-        self.shutdown_worker()
-
-    def log_diagnostics(self, paths):
-        self.policy.log_diagnostics(paths)
-        self.baseline.log_diagnostics(paths)
-
-    def init_opt(self):
-        """
-        Initialize the optimization procedure. If using tf, this may
-        include declaring all the variables and compiling functions
-        """
-        raise NotImplementedError
+        self.discount = discount
+        self.max_path_length = max_path_length
 
-    def get_itr_snapshot(self, itr, samples_data):
-        """
-        Returns all the data that should be saved in the snapshot for this
-        iteration.
+        self.episode_reward_mean = deque(maxlen=100)
+
+    def process_samples(self, itr, paths):
+        """Return processed sample data based on the collected paths.
+        This method is currently used by cem and cam_es.
+
+        Args:
+            itr (int): Iteration number.
+            paths (list[dict]): A list of collected paths
+
+        Returns:
+            dict: Processed sample data, with key
+                * average_return: (float)
         """
-        raise NotImplementedError
+        baselines = []
+        returns = []
+
+        if hasattr(self.baseline, 'predict_n'):
+            all_path_baselines = self.baseline.predict_n(paths)
+        else:
+            all_path_baselines = [
+                self.baseline.predict(path) for path in paths
+            ]
+
+        for idx, path in enumerate(paths):
+            # baselines
+            path['baselines'] = all_path_baselines[idx]
+            baselines.append(path['baselines'])
+
+            # returns
+            path['returns'] = special.discount_cumsum(path['rewards'],
+                                                      self.discount)
+            returns.append(path['returns'])
+
+        undiscounted_returns = [sum(path['rewards']) for path in paths]
+
+        samples_data = dict(average_return=np.mean(undiscounted_returns))
+
+        tabular.record('Iteration', itr)
+        tabular.record('AverageReturn', np.mean(undiscounted_returns))
```

Comment:
all of the metrics logged in tf.algos.BatchPolopt (StdReturn, etc.) are useful and should still be logged here.

Reply:
```
ent = np.sum(self.policy.distribution.entropy(agent_infos) *
                     valids) / np.sum(valids)
```
> if you only return average_return, why calculate all this other stuff?

I kept all the logging used variables last time. Would you tell me which one is not necessary to calculate?
```
        tabular.record('Iteration', itr)
        tabular.record('AverageDiscountedReturn', average_discounted_return)
        tabular.record('AverageReturn', np.mean(undiscounted_returns))
        tabular.record('Extras/EpisodeRewardMean',
                       np.mean(self.episode_reward_mean))
        tabular.record('NumTrajs', len(paths))
        tabular.record('Entropy', ent)
        tabular.record('Perplexity', np.exp(ent))
        tabular.record('StdReturn', np.std(undiscounted_returns))
        tabular.record('MaxReturn', np.max(undiscounted_returns))
        tabular.record('MinReturn', np.min(undiscounted_returns))

```

Reply:
If you mean tf.algos.BatchPolopt, it does have all the things. For np it seems only AverageReturn is needed.

Reply:
i guess what i'm saying is that all on-policy algorithms need to log these statistics, so in that sense, many of these calculations are "needed," even if they aren't used during calculations in the algorithms.

Reply:
let's keep all of them for now. just file an issue to extract the common rollout statistics into a library, rather than copy/pasting them between classes